Sources/SwiftBSON/BSONDocument.swift

```diff
@@ -74,12 +74,27 @@ public struct BSONDocument {
      * - SeeAlso: http://bsonspec.org/
      */
     public init(fromBSON bson: ByteBuffer) throws {
-        let storage = BSONDocumentStorage(bson)
+        let storage = BSONDocumentStorage(bson.slice())
         try storage.validate()
-        self = BSONDocument(fromUnsafeBSON: storage)
+        self.storage = storage
+    }
+
+    /**
+     * Initialize a new `BSONDocument` from the provided BSON data without validating the elements. The first four
+     * bytes must accurately reflect the length of the buffer, however.
+     *
+     * If invalid BSON data is provided, undefined behavior or server-side errors may occur when using the
+     * resultant `BSONDocument`.
+     *
+     * - Throws: `BSONError.InvalidArgumentError` if the provided BSON's length does not match the encoded length.
+     */
+    public init(fromBSONWithoutValidatingElements bson: ByteBuffer) throws {
```

Comment:
One thing that isn't reflected in this PR is that something like the following _could_ panic:
```swift
let doc = try BSONDocument(fromBSONWithoutValidatingElements: bson)
for i in 0..<doc.count {
    print(doc[i])
}
```
This is because the index-based subscript doesn't return an optional since it just `fatalErrors` when an out-of-bounds index is provided. Given that this subscript already can `fatalError` pretty easily, it seemed okay to me to `fatalError` on invalid BSON too, even if `count` says its okay. What do you think?